### PR TITLE
Start a vote on require(esm) markers

### PR DESCRIPTION
**Pull request by joyeecheung**
This is a follow-up to the naming bikeshed in https://github.com/nodejs/loaders/issues/221 which is for the implementation in https://github.com/nodejs/node/pull/54563 A summary can be found in https://github.com/nodejs/node/pull/54563#pullrequestreview-2288424604

### CODE REVIEW

**Comment by joyeecheung** [comment]
cc @nodejs/tsc @nodejs/loaders @guybedford

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [inline votes/initiateNewVote/_EDIT_ME.yml:40]:
```suggestion
  This is a follow-up to the naming bikeshed in https://github.com/nodejs/loaders/issues/221
  which is for the implementation in https://github.com/nodejs/node/pull/54563
  and the follow-on issue in https://github.com/nodejs/node/pull/53848
  A short summary can be found in https://github.com/nodejs/node/pull/54563#pullrequestreview-2288424604
```

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [inline votes/initiateNewVote/_EDIT_ME.yml:83]: I'd like to include a description of the follow-on work here;

```suggestion
  As a follow-up feature, on the other side, when importing some CJS module `bar` into ESM:
  
  bar.js
  ```js
    module.exports = 'bar';
  ```js
  
  Users importing with `console.log(await import('bar'))` the user will get a namespace object of the form `ModuleNamespace { default: 'bar' }`.
  
  To support stable transitive interop patterns that can work both ways, the same flag can be exported on the namespace object when CJS modules are represented in ESM - either `ModuleNamespace { __cjsUnwrapDefault: true, default: 'bar' }` or `ModuleNamespace { default: 'bar', 'module.exports': 'bar' }` depending on which option is chosen for this first PR.
```

Note the closing code sections need to have the js suffix removed.

**Review by guybedford — CHANGES_REQUESTED** [review]
Thanks @joyeecheung for putting this together. Just added some extra notes on the add-on PR here, as it is also part of the implications here.

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (joyeecheung) [reply votes/initiateNewVote/_EDIT_ME.yml:83]: I don't think we have consensus about using the namespace for this, and that might also come down to a vote, but I can leave a link about the potential followup.

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (aduh95) [inline votes/initiateNewVote/_EDIT_ME.yml:6]:
```suggestion
subject: Choosing a marker to customize default exports returned by `require(esm)`
```

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (aduh95) [inline votes/initiateNewVote/_EDIT_ME.yml:28]:
```suggestion
  - '`export { foo as "module.exports" };`'
  - '`export const __cjsUnwrapDefault = true;`'
```

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [reply votes/initiateNewVote/_EDIT_ME.yml:84]: @joyeecheung when we first discussed this over two months ago now you explicitly told me you were okay with this, and now you are saying you might not only hinder this first phase compromise but potentially also block the second part of my PR. We need to do better as collaborators to not waste eachother's time.

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [inline votes/initiateNewVote/_EDIT_ME.yml:87]: This is not resolved, and my request for including the explicit shape of the namespace remains as per my comment.

**Review by guybedford — CHANGES_REQUESTED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (joyeecheung) [reply votes/initiateNewVote/_EDIT_ME.yml:84]: If I am looking at the timelines correctly, it went like this:

1. I was okay with it at first, because I thought it was not too much too add
2. I then realized that this would be semver-major since it would break the namespace, which I had not really realized before, so the "not too much to add" is no longer valid. And then, I realized we could potentially solve the use case without breaking the namespace with e.g. https://github.com/nodejs/node/pull/53848#issuecomment-2244500584

I think in this project, we are allowed to change our minds when we have new discoveries about the implications of a proposal. At the end of the day, when there are disagreements, we go to a vote. I am fine with breaking the namespace if the majority of the group supports it, but I don't feel comfortable to say "I am fine with it because I had said so when I did not realize there were more implications & alternatives, and I would just pretend that I didn't have any of the realizations to speed this up, and let the decision be made by just these two people".

Sorry if you feel that your time has been wasted. I believe we are just trying to make module loading better in Node.js, even though we have different priorities.

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (joyeecheung) [reply votes/initiateNewVote/_EDIT_ME.yml:87]: Isn't the referenced link https://github.com/nodejs/node/pull/53848 already including that proposed namespace?

**Review by ShogunPanda — APPROVED** [review]

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [reply votes/initiateNewVote/_EDIT_ME.yml:84]: The issue is that you've put me in a situation where I need a vote on PR to land my compromise to then land my other compromise of a major change (with the delays now causing us to miss the 23 release milestone).

There is a very real impact on progress that comes with this behaviour, and while you have every right to object, I have every right to feel annoyed!

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [reply votes/initiateNewVote/_EDIT_ME.yml:87]: The wording here is really dismissive to the use case I've been working on for two months @joyeecheung which is the reason I'm working on all this, which I really don't appreciate.

Transitive interop patterns are important and a key feature of this flag for the future of transpilation of CommonJS into ESM. The vote description absolutely needs to cover this aspect and the use case I'm working on here.

Yes we have different use cases, and I've compromised my use cases to solve yours, but you cannot go around dismissing my use case discussions as abstract hypothetical future scenarios, it's plainly rude as written here.

To correct this line:

```suggestion
  Further to this vote, there is [a follow-on proposal](https://github.com/nodejs/node/pull/53848) to reuse this marker on namespaces representing CJS modules imported into ESM, where attaching this marker would effectively define a scheme for converting CJS into ESM that would retain full transitive interop semantics allowing arbitrary CJS to be converted into ESM as a non-breaking change.
```

My suggestion remains:

```js
Under this follow-on marking, when importing some CJS module `bar` into ESM:
  
  bar.js
  ```js
    module.exports = 'bar';
  ```js
  
  Users importing with `console.log(await import('bar'))` the user would a namespace object either of the form `ModuleNamespace { __cjsUnwrapDefault: true, default: 'bar' }` or `ModuleNamespace { default: 'bar', 'module.exports': 'bar' }` depending on which option is chosen for this first PR.
```

Comment (joyeecheung) [reply votes/initiateNewVote/_EDIT_ME.yml:87]: I am not sure what part about the wording is rude, or dismissive. I am just trying to keep the description on-point and leave the additional context to a link.

>  Note that to follow up with this vote, there is [a proposal](https://github.com/nodejs/node/pull/53848) to reuse this marker for detecting whether the imported module is CJS.

I think this is just explaining the situation as it is, without derailing the vote with another topic.

>  There is no consensus on whether that is the right way to solve the problem yet ([see](https://github.com/nodejs/node/pull/53848#issuecomment-2244500584)), so this vote is just about
  the marker naming bikeshed.

I think this is just describing the fact? We don't have consensus that the reuse is the right way for `import(cjs)` detection, and this vote is not about that, either?

But anyway, if you insist on including the additional information and making the text longer by about 800+ words, I can include them. Personally I think the longer the text is, the harder it is for voters to digest, or at least when I am asked to vote, I would hope that the summary can just cut to the chase and keep the additional context in links. But there is no need to let the vote get stuck here.

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (joyeecheung) [reply votes/initiateNewVote/_EDIT_ME.yml:84]: Technically, I didn't formally block the PRs (with request changes), which was also what I said 2 months ago - I would not block them but I was also skeptical about them - so when you have enough approvals after the final commit, you could land them despite my concerns. If I didn't want progress to be made, I would not suggest to vote in the first place - the TSC vote is specifically used to prevent a standstill and come to a decision.

I understand that maybe in TC39 or the former modules group at some point, there is a "one person disagrees and it will never proceed" situation, and that might lead you to believe that Node.js had the same model and you absolutely needed my approval to land the PRs. But no, in Node.js, you could either use a vote to accelerate the decision making, or just land them despite having non-blocking concerns at your own discretion.

File: votes/initiateNewVote/_EDIT_ME.yml
Comment (guybedford) [inline votes/initiateNewVote/_EDIT_ME.yml:93]: Minor typo from my suggestion before:

```suggestion
  Users importing with `console.log(await import('bar'))` would obtain a namespace object either of the form `ModuleNamespace { __cjsUnwrapDefault: true, default: 'bar' }` or `ModuleNamespace { default: 'bar', 'module.exports': 'bar' }` depending on which option is chosen for this first PR.
```

**Review by guybedford — APPROVED** [review]
Thanks @joyeecheung for taking on board the feedback, happy to go ahead, and as I say I completely commit to following whatever is voted on here.

**Review by mhdawson — APPROVED** [review]
LGTM

**Comment by joyeecheung** [comment]
I am not entirely sure how long this should wait, https://github.com/nodejs/TSC/tree/main/votes/initiateNewVote#initiate-a-new-tsc-vote just says "Once the PR has approvals", and it's been open for about 2 days with a few approvals, so I guess I'll just merge it when it passes the 48-hour mark.

**Review by marco-ippolito — APPROVED** [review]

**Review by MoLow — APPROVED** [review]

**Pull request merged**
